source/MaterialXView/Viewer.cpp

```diff
@@ -1621,31 +1672,60 @@ void Viewer::bakeTextures()
     MaterialPtr material = getSelectedMaterial();
     mx::ShaderRefPtr shaderRef = material->getElement()->asA<mx::ShaderRef>();
     mx::FileSearchPath searchPath = _searchPath;
-    if (material->getDocument())
+    mx::DocumentPtr doc = material->getDocument();
+    if (!doc)
     {
-        mx::FilePath documentFilename = material->getDocument()->getSourceUri();
-        searchPath.append(documentFilename.getParentPath());
+        return;
     }
-
+    mx::FilePath documentFilename = doc->getSourceUri();
+    searchPath.append(documentFilename.getParentPath());
+    mx::ValuePtr udimSetValue = doc->getGeomPropValue("udimset");
+    mx::Image::BaseType baseType = (_bakeHdr)? mx::Image::BaseType::FLOAT : mx::Image::BaseType::UINT8;
     mx::ImageHandlerPtr imageHandler = mx::GLTextureHandler::create(mx::StbImageLoader::create());
     imageHandler->setSearchPath(searchPath);
-    if (!material->getUdim().empty())
-    {
-        mx::StringResolverPtr resolver = mx::StringResolver::create();
-        resolver->setUdimString(material->getUdim());
-        imageHandler->setFilenameResolver(resolver);
-    }
 
-    try
+    // if material has udims
+    if (!material->getUdim().empty() && udimSetValue && _materials.size() > 1)
     {
-        mx::TextureBakerPtr baker = mx::TextureBaker::create();
-        baker->setImageHandler(imageHandler);
-        baker->bakeShaderInputs(shaderRef, _genContext, _bakeFilename.getParentPath());
-        baker->writeBakedDocument(shaderRef, _bakeFilename);
+        for (MaterialPtr mat : _materials)
+        {
+            if (mat->getElement()->getCategory() == "shaderref")
+            {
+                mx::StringResolverPtr resolver = mx::StringResolver::create();
+                resolver->setUdimString(mat->getUdim());
+                imageHandler->setFilenameResolver(resolver);
+
+                try
+                {
+                    mx::TextureBakerPtr baker = mx::TextureBaker::create(_bakeTextureRes, _bakeTextureRes, baseType);
+                    baker->setImageHandler(imageHandler);
+                    baker->bakeShaderInputs(shaderRef, _genContext, _bakeFilename.getParentPath(), mat->getUdim());
+                    std::vector<std::string> udimSet = udimSetValue->asA<std::vector<std::string>>();
+                    if (mat->getUdim() == udimSet.back())
+                    {
+                        baker->writeBakedDocument(shaderRef, _bakeFilename, udimSetValue);
```

Comment:
Would it be possible for the viewer to leverage bakeAllShaders as well, or are there different requirements for this version of the code?

Reply:
The Viewer chooses the selected material and leverages the existing imagehandler presets in the viewer. BakeAllShaders just uses the explicitly stated material in the document and creates a basic default imagehandler from scratch. I would say BakeAllShaders is a little more naive compared to what is done in Viewer.cpp.

Reply:
Got it, and that all makes sense.